src/mds/PurgeQueue.cc

```diff
@@ -0,0 +1,284 @@
+
+    return;
+  }
+
+  if (!journaler.is_readable()) {
+    dout(10) << " not readable right now" << dendl;
+    if (!journaler.have_waiter()) {
+      journaler.wait_for_readable(new FunctionContext([this](int r) {
+        Mutex::Locker l(lock);
+        if (r == 0) {
+          _consume();
+        }
+      }));
+    }
+
+    return;
+  }
+
+  // The journaler is readable: consume an entry
+  bufferlist bl;
+  bool readable = journaler.try_read_entry(bl);
+  assert(readable);  // we checked earlier
+
+  dout(20) << " decoding entry" << dendl;
+  PurgeItem item;
+  bufferlist::iterator q = bl.begin();
+  ::decode(item, q);
+  dout(20) << " executing item (0x" << std::hex << item.ino
+           << std::dec << ")" << dendl;
+  _execute_item(item, journaler.get_read_pos());
+}
+
+void PurgeQueue::_execute_item(
+    const PurgeItem &item,
+    uint64_t expire_to)
+{
+  assert(lock.is_locked_by_me());
+
+  in_flight[expire_to] = item;
+
+  // TODO: handle things other than normal file purges
+  // (directories, snapshot truncates)
+  C_GatherBuilder gather(cct);
+  if (item.size > 0) {
+    uint64_t num = Striper::get_num_objects(item.layout, item.size);
+    dout(10) << " 0~" << item.size << " objects 0~" << num
+             << " snapc " << item.snapc << " on " << item.ino << dendl;
+    filer.purge_range(item.ino, &item.layout, item.snapc,
+                      0, num, ceph::real_clock::now(), 0,
+                      gather.new_sub());
+  }
+
+  // remove the backtrace object if it was not purged
+  object_t oid = CInode::get_object_name(item.ino, frag_t(), "");
+  if (!gather.has_subs() || !item.layout.pool_ns.empty()) {
+    object_locator_t oloc(item.layout.pool_id);
+    dout(10) << " remove backtrace object " << oid
+	     << " pool " << oloc.pool << " snapc " << item.snapc << dendl;
+    objecter->remove(oid, oloc, item.snapc,
+			  ceph::real_clock::now(), 0,
+			  NULL, gather.new_sub());
+  }
+
+  // remove old backtrace objects
+  for (const auto &p : item.old_pools) {
+    object_locator_t oloc(p);
+    dout(10) << " remove backtrace object " << oid
+	     << " old pool " << p << " snapc " << item.snapc << dendl;
+    objecter->remove(oid, oloc, item.snapc,
+			  ceph::real_clock::now(), 0,
+			  NULL, gather.new_sub());
+  }
+  assert(gather.has_subs());
+
+  gather.set_finisher(new FunctionContext([this, expire_to](int r){
+    execute_item_complete(expire_to);
+  }));
+  gather.activate();
+}
+
+void PurgeQueue::execute_item_complete(
+    uint64_t expire_to)
+{
+  dout(10) << "complete at 0x" << std::hex << expire_to << std::dec << dendl;
+  Mutex::Locker l(lock);
+  assert(in_flight.count(expire_to) == 1);
+
+  auto iter = in_flight.find(expire_to);
+  assert(iter != in_flight.end());
+  if (iter == in_flight.begin()) {
+    // This was the lowest journal position in flight, so we can now
+    // safely expire the journal up to here.
+    journaler.set_expire_pos(expire_to);
```

Comment:
The expire_to value passed into _execute_item from _consume is the journal read pos *after* the entry has been read, i.e. the next entry -- so I think we're already doing this